Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/features/input/model/DolphinSensorEventListener.java

```diff
@@ -0,0 +1,434 @@
+{
+  // Set of three axes. Creates a negative companion to each axis, and corrects for device rotation.
+  private static final int AXIS_SET_TYPE_DEVICE_COORDINATES = 0;
+  // Set of three axes. Creates a negative companion to each axis.
+  private static final int AXIS_SET_TYPE_OTHER_COORDINATES = 1;
+
+  private static class AxisSetDetails
+  {
+    public final int firstAxisOfSet;
+    public final int axisSetType;
+
+    public AxisSetDetails(int firstAxisOfSet, int axisSetType)
+    {
+      this.firstAxisOfSet = firstAxisOfSet;
+      this.axisSetType = axisSetType;
+    }
+  }
+
+  private static class SensorDetails
+  {
+    public final int sensorType;
+    public final String[] axisNames;
+    public final AxisSetDetails[] axisSetDetails;
+    public boolean isSuspended = true;
+
+    public SensorDetails(int sensorType, String[] axisNames, AxisSetDetails[] axisSetDetails)
+    {
+      this.sensorType = sensorType;
+      this.axisNames = axisNames;
+      this.axisSetDetails = axisSetDetails;
+    }
+  }
+
+  private static int sDeviceRotation = Surface.ROTATION_0;
+
+  private final SensorManager mSensorManager;
+
+  private final HashMap<Sensor, SensorDetails> mSensorDetails = new HashMap<>();
+
+  private final boolean mRotateCoordinatesForScreenOrientation;
+
+  private String mDeviceQualifier = "";
+
+  // The fastest sampling rate Android lets us use without declaring the HIGH_SAMPLING_RATE_SENSORS
+  // permission is 200 Hz. This is also the sampling rate of a Wii Remote, so it fits us perfectly.
+  private static final int SAMPLING_PERIOD_US = 1000000 / 200;
+
+  @Keep
+  public DolphinSensorEventListener()
+  {
+    mSensorManager = (SensorManager)
+            DolphinApplication.getAppContext().getSystemService(Context.SENSOR_SERVICE);
+    mRotateCoordinatesForScreenOrientation = true;
+
+    addSensors();
+  }
+
+  @Keep
+  public DolphinSensorEventListener(InputDevice inputDevice)
+  {
+    mRotateCoordinatesForScreenOrientation = false;
+
+    if (Build.VERSION.SDK_INT >= 31)
+    {
+      mSensorManager = inputDevice.getSensorManager();
+
+      addSensors();
+    }
+    else
+    {
+      mSensorManager = null;
+    }
+  }
+
+  private void addSensors()
+  {
+    tryAddSensor(Sensor.TYPE_ACCELEROMETER, new String[]{"Accel Right", "Accel Left",
+                    "Accel Forward", "Accel Backward", "Accel Up", "Accel Down"},
+            new AxisSetDetails[]{new AxisSetDetails(0, AXIS_SET_TYPE_DEVICE_COORDINATES)});
+
+    tryAddSensor(Sensor.TYPE_GYROSCOPE, new String[]{"Gyro Pitch Up", "Gyro Pitch Down",
+                    "Gyro Roll Right", "Gyro Roll Left", "Gyro Yaw Left", "Gyro Yaw Right"},
+            new AxisSetDetails[]{new AxisSetDetails(0, AXIS_SET_TYPE_DEVICE_COORDINATES)});
+
+    tryAddSensor(Sensor.TYPE_LIGHT, "Light");
+
+    tryAddSensor(Sensor.TYPE_PRESSURE, "Pressure");
+
+    tryAddSensor(Sensor.TYPE_TEMPERATURE, "Device Temperature");
```

Comment:
```suggestion
    tryAddSensor(Sensor.TYPE_AMBIENT_TEMPERATURE, "Device Temperature");
```

Is there a reason we shouldn't use the supported ambient temperature sensor constant?

Reply:
That one is also in the list, later on.

Reply:
Oh sorry, but could that add the same sensor twice? Admittedly this is a silly sensor to bug you on. I'm just a little confused.

Reply:
I thought `TYPE_TEMPERATURE` was for the temperature inside the device. I could be wrong though.

Reply:
`TYPE_TEMPERATURE` was replaced by `TYPE_AMBIENT_TEMPERATURE`

Reply:
Yes, but they are not the exact same thing.

Reply:
Huh, that's a terrible android framework thing then... I'll let this one go then.